src/execCmd.ts

```diff
@@ -0,0 +1,210 @@
+
+  if (isNumber(cmdOptions.ensureExitCode) && result.shellOutput.code !== cmdOptions.ensureExitCode) {
+    throw getExitCodeError(cmdOptions.ensureExitCode, result.shellOutput.code, cmd);
+  }
+
+  return addJsonOutput(cmd, result);
+};
+
+/**
+ * Asynchronously execute a command with the provided options in a child process.
+ *
+ * Option defaults:
+ *    1. `cwd` = process.cwd()
+ *    2. `timeout` = 300000 (5 minutes)
+ *    3. `env` = process.env
+ *    4. `silent` = true (child process output not written to the console)
+ *
+ * Other defaults:
+ *
+ *    @see www.npmjs.com/package/shelljs#execcommand--options--callback
+ *    @see www.nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback
+ *
+ * @param cmd The command string to be executed by a child process.
+ * @param options The options used to run the command.
+ * @returns The child process exit code, stdout, stderr, cmd run time, and the parsed JSON if `--json` param present.
+ */
+export const execCmdAsync = async (cmd: string, options: ExecCmdOptions = {}): Promise<ExecCmdResult> => {
+  const debug = Debug('testkit:execCmdAsync');
+
+  const resultPromise = new Promise<ExecCmdResult>((resolve, reject) => {
+    // Ensure we run asynchronously
+    if (options.async === false) {
+      reject(new Error('execCmdAsync must be run asynchronously.  Use execCmd to run synchronously.'));
+    }
+
+    const cmdOptions = Object.assign({}, DEFAULT_SHELL_OPTIONS, options);
+
+    debug(`Running cmd: ${cmd}`);
+    debug(`Cmd options: ${inspect(cmdOptions)}`);
+
+    const callback: ExecCallback = (code, stdout, stderr) => {
+      const execCmdDuration = hrtimeToMillisDuration(process.hrtime(startTime));
+      debug(`Command completed with exit code: ${code}`);
+
+      if (isNumber(cmdOptions.ensureExitCode) && code !== cmdOptions.ensureExitCode) {
+        reject(getExitCodeError(cmdOptions.ensureExitCode, code, cmd));
+      }
+
+      const result: ExecCmdResult = {
+        shellOutput: new ShellString(stdout),
+        execCmdDuration,
+      };
+      result.shellOutput.code = code;
+      result.shellOutput.stdout = stdout;
+      result.shellOutput.stderr = stderr;
+
+      resolve(addJsonOutput(cmd, result));
+    };
+
+    // Execute the command async in a child process
+    const startTime = process.hrtime();
+    shelljs.exec(cmd, cmdOptions, callback);
+  });
+
+  return resultPromise;
+};
+
+/**
+ * Build a command string using an optional binary path for use by
+ * execCmd or execCmdAsync.
+ *
+ * The binary preference order is:
+ *    1. binaryPath arg
+ *    2. TESTKIT_BINARY_PATH env var
+ *    3. `bin/run` (default)
+ *
+ * @param cmdArgs The command name and args as a string. E.g., `"force:user:create -a testuser1"`
+ * @param binaryPath The path to a command executable. E.g., `"node_modules/bin/sfdx"`
+ * @returns The command string with CLI binary. E.g., `"node_modules/bin/sfdx force:user:create -a testuser1"`
```

Comment:
would it be safer to return a full, non-relative path?  I'm thinking of all the permutations we may end up with (lerna monorepos, test projects inside a plugin, all the things people may try to do with cwd including our mock of it)

Reply:
I could be wrong but I'm not sure it matters.  When I think about the use of this function I anticipate this is called just before an execCmd* function, in which case you are already dealing with cwd and where the command will be executed.  There's nothing preventing anyone from passing in a fully resolved path to the binary if they want, and that way we won't need to deal with more special processing if `sfdx` is passed as a path.

Reply:
@shetzel To Shane's point of project complexity, maybe it would be useful if the debug for `Using binary` would add the full path to the binary, in addition to the calculated value. Might help untangle issues when projects get complicated.

Reply:
Ok, I'll return the resolved path and enhance debug output

Reply:
@shetzel All I think needs to be done is the enhanced debug output. Leaving the path as the caller specified seems like the right thing to do.